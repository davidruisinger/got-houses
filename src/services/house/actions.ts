import { createAction } from 'typesafe-actions'
import { House } from './types'

/**
 * Redux action to fetch houses.
 * @param {Object} config
 * @param {number} config.page - The page that should be fetched
 * @param {number} config.pageSize - The size of one page
 */
export const fetchHouses = createAction('GET_HOUSES_REQUEST', resolve => {
  return ({ pageSize = 10 }: { pageSize: number }) => resolve({ pageSize })
})

/**
 * Redux action to fetch a single house by ID.
 * @param {Object} config
 * @param {number} config.id - The ID of the house
 */
export const fetchHouseById = createAction(
  'GET_HOUSE_BY_ID_REQUEST',
  resolve => {
    return ({ id }: { id: string }) => resolve({ id })
  }
)

/**
 * Redux action to process an array of fetched houses.
 * @param {Object} result
 * @param {string} result.houses - The fetched houses.
 */
export const fetchHousesSuccess = createAction(
  'GET_HOUSES_SUCCESS',
  resolve => {
    return ({
      houses,
      isMoreAvailable,
    }: {
      houses: House[]
      isMoreAvailable: boolean | 'skip' // If skip is provided the reducer will ignore this (used for single fetches)
    }) => resolve({ houses, isMoreAvailable })
  }
)

/**
 * Redux action to process an error while fetching houses.
 * @param {Object} error
 * @param {string} error.message - The error message.
 */
export const fetchHousesFail = createAction('GET_HOUSES_FAIL', resolve => {
  return ({ message }: { message: string }) => resolve({ message })
})
