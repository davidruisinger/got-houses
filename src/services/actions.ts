// We do not really need this except for generating the global ActionType
import * as houseActions from './house/actions'

export default {
  house: houseActions,
}
