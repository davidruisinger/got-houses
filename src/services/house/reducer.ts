import { ActionType, getType } from 'typesafe-actions'
import * as actions from './actions'
import { House } from './types'

export type Houses = { [id: string]: House }

export type HouseState = Readonly<{
  houses: Houses
  isFetching: boolean
  isMoreAvailable: boolean
  error: null | string
}>

const initialState: HouseState = {
  houses: {},
  isFetching: false,
  isMoreAvailable: true,
  error: null,
}

export default (
  state: HouseState = initialState,
  action: ActionType<typeof actions>
): HouseState => {
  switch (action.type) {
    case getType(actions.fetchHouses):
    case getType(actions.fetchHouseById):
      return {
        ...state,
        isFetching: true,
        error: null,
      }

    case getType(actions.fetchHousesSuccess):
      // Normalize the array to an object
      const normalizedHouses: {
        [id: string]: House
      } = action.payload.houses.reduce((obj: Houses, house) => {
        // There is no ID field so we need to get the ID from the url
        const id = house.url.split('/').pop()
        // We also transform the URL for the potentially set overlord into an ID
        const overlordId = house.overlord.split('/').pop() || ''
        if (id) obj[id] = { ...house, id, overlordId }

        return obj
      }, {})
      const isMoreAvailable = action.payload.isMoreAvailable

      return {
        ...state,
        houses: {
          ...state.houses,
          ...normalizedHouses,
        },
        isMoreAvailable:
          isMoreAvailable == 'skip' ? state.isMoreAvailable : isMoreAvailable,
        isFetching: false,
      }

    case getType(actions.fetchHousesFail):
      return {
        ...state,
        isFetching: false,
        error: action.payload.message,
      }

    default:
      return state
  }
}
