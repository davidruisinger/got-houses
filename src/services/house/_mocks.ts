import { House } from './types'
import { HouseState } from './reducer'

export const housesMock: House[] = [
  {
    id: '1',
    url: 'https://www.anapioficeandfire.com/api/houses/1',
    name: 'Test House',
    region: 'Test Region',
    coatOfArms: 'Something',
    words: 'foo baar',
    titles: ['The Testers'],
    seats: ['Test seat'],
    currentLord: 'https://www.anapioficeandfire.com/api/characters/1',
    heir: 'https://www.anapioficeandfire.com/api/characters/1',
    overlord: 'https://www.anapioficeandfire.com/api/characters/1',
    overlordId: '1',
    founded: '',
    founder: 'https://www.anapioficeandfire.com/api/characters/1',
    diedOut: '',
    ancestralWeapons: ['Test Sword'],
    cadetBranches: [],
    swornMembers: ['https://www.anapioficeandfire.com/api/characters/1'],
  },
]

export const houseStateMock: HouseState = {
  isFetching: false,
  error: null,
  isMoreAvailable: true,
  houses: {
    [housesMock[0].name]: housesMock[0],
  },
}
