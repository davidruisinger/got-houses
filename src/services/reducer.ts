import { combineReducers } from 'redux'

// Import each service reducer and it's name
import { SERVICE_NAME as houseServiceName } from './house/constants'
import HouseReducer from './house/reducer'

const rootReducer = combineReducers({
  [houseServiceName]: HouseReducer,
})

export default rootReducer
