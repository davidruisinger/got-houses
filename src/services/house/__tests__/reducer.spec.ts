import reducer, { HouseState } from '../reducer'
import * as actions from '../actions'
import { housesMock } from '../_mocks'

const initialState = reducer(undefined, {} as any)

describe('house service reducer', () => {
  describe('initial state', () => {
    it('should match the snapshot', () => {
      expect(initialState).toMatchSnapshot()
    })
  })

  let fetchingState: HouseState,
    successState: HouseState,
    errorState: HouseState
  describe('fetchHouses action', () => {
    it('should set isFetching', () => {
      const action = actions.fetchHouses({
        pageSize: 4,
      })
      fetchingState = reducer(initialState, action)
      expect(fetchingState.isFetching).toBeTruthy()
    })
  })

  describe('fetchHouseById action', () => {
    it('should set isFetching', () => {
      const action = actions.fetchHouseById({
        id: '1',
      })
      fetchingState = reducer(initialState, action)
      expect(fetchingState.isFetching).toBeTruthy()
    })
  })

  describe('fetchHousesSuccess action', () => {
    it('should unset isFetching and set the houses', () => {
      const action = actions.fetchHousesSuccess({
        houses: housesMock,
        isMoreAvailable: true,
      })
      successState = reducer(fetchingState, action)
      expect(successState.isFetching).toBeFalsy()
      const id = housesMock[0].url.split('/').pop()
      expect(successState.houses).toEqual({
        // @ts-ignore
        [id]: housesMock[0],
      })
    })

    it('should only add new houses to an existing list of houses', () => {
      const modifiedHouse = {
        ...housesMock[0],
        url: 'https://www.anapioficeandfire.com/api/houses/2',
        id: '2',
      }
      const modifiedHousesMocks = [
        housesMock[0], // This should not be added a second time
        modifiedHouse,
      ]
      const action = actions.fetchHousesSuccess({
        houses: modifiedHousesMocks,
        isMoreAvailable: true,
      })
      successState = reducer(successState, action)
      expect(Object.keys(successState.houses)).toHaveLength(2)
      const id = housesMock[0].url.split('/').pop()
      // @ts-ignore
      expect(successState.houses[id]).toEqual(housesMock[0])
      expect(successState.houses['2']).toEqual(modifiedHouse)
    })

    it('should skip setting isMoreAvailable if requested', () => {
      expect(successState.isMoreAvailable).toBeTruthy()
      const action = actions.fetchHousesSuccess({
        houses: [],
        isMoreAvailable: 'skip',
      })
      successState = reducer(successState, action)
      expect(successState.isMoreAvailable).toBeTruthy()
    })

    it('should set isMoreAvailable to false when there is nothing more to fetch', () => {
      expect(successState.isMoreAvailable).toBeTruthy()
      const action = actions.fetchHousesSuccess({
        houses: [],
        isMoreAvailable: false,
      })
      successState = reducer(successState, action)
      expect(successState.isMoreAvailable).toBeFalsy()
    })
  })

  describe('fetchHousesFail action', () => {
    it('should unset isFetching and set an error', () => {
      const errorMessage = 'Test error'
      const action = actions.fetchHousesFail({ message: 'Test error' })
      errorState = reducer(fetchingState, action)
      expect(errorState.isFetching).toBeFalsy()
      expect(errorState.error).toBe(errorMessage)
    })
  })

  describe('fetchHouses action', () => {
    it('should set isFetching and clear the error when re-trying', () => {
      const action = actions.fetchHouses({
        pageSize: 4,
      })
      const retryState = reducer(errorState, action)
      expect(retryState.isFetching).toBeTruthy()
      expect(retryState.error).toBe(null)
    })
  })
})
