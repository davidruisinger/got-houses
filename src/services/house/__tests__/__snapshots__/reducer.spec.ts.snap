// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`house service reducer initial state should match the snapshot 1`] = `
Object {
  "error": null,
  "houses": Object {},
  "isFetching": false,
  "isMoreAvailable": true,
}
`;
